### ISSUE TITLE
Particles crash in Firefox

### ISSUE THREAD

**Issue by heniotierra** [craftyjs/Crafty#847]
Particles in Firefox are very slow and in some machines may cause the browser to crash. It's bad in Linux, but seems to be much worse in Windows.

I'm using the settings below to create the particles. Maybe the problem is in some of the parameters?

```
var particlesOptions = {
        maxParticles: 200,
        size: 5,
        sizeRandom: 4,
        speed: .1,
        speedRandom: .3,
        lifeSpan: 70,
        lifeSpanRandom: 10,
        angle: 225,
        angleRandom: 35,
        startColour: [0, 204, 255, 1],
        startColourRandom: [60, 60, 60, .2],
        endColour: [0, 255, 255, 0],
        endColourRandom: [60, 60, 60, .2],
        sharpness: 20,
        sharpnessRandom: 10,
        spread: Crafty.viewport.width,
        duration: -1,
        fastMode: false,
        gravity: { x: -0.05, y: 0.1 },
        jitter: 3
};
```

**Comment by heniotierra** [comment]
I found that turning shapness off and fastMode to true, as well as lowering max particles, helps a lot, but still it's considerably slower than in Chrome. I wonder why there's such huge difference between Firefox's perfomance and Chrome's with particles.

**Comment by starwed** [comment]
Presumably Chrome just has a more efficient implementation of `fillRect` than Firefox?  I think Chrome is generally much smoother/faster than Firefox when it comes to Canvas.

Rendering without `fastMode` looks like it does some fancy radial gradients, so I'm not surprised that's (in general) much slower.  (And I believe sharpness is only used for some calculations with radial gradients.)

The particle rendering code is pretty ancient at this point, and tbh I haven't ever analyzed exactly what it's doing to see if there are ways to speed it up.  Might be interesting to see if a version using WebGL could be made.

**Comment by starwed** [comment]
Looking through the code, I see that it does several quasi-expensive operations for _each_ particle _every_ frame.  (Creates some arrays/objects in the inner loop, for instance.)

This could definitely be made more light weight, but it could also be completely rewritten.

I saw [this post](http://www.reddit.com/r/gamedev/comments/2vlypg/i_made_a_html5_particle_engine/) on reddit; it apparently gets pretty good performance.  I poked the OP about what sort of license they'd release it under.

**Comment by mucaho** [comment]
In the meantime we could improve performance by handling transformations by the canvas matrices, rather than transforming particles individually.
see [old particle PR](https://github.com/craftyjs/Crafty/pull/743#issuecomment-36547024)

**Comment by mucaho** [comment]
should be fixed by #1132
